src/uri.rs

```diff
@@ -1,11 +1,260 @@
+            }
+            let query_len = url.query().unwrap_or("").len();
+            let authority_end = s.split(url.path()).next().unwrap_or(s).len();
+            let fragment_len = url.fragment().unwrap_or("").len();
+            match url.origin() {
+                url::Origin::Opaque(_) => Err(Error::Method),
+                url::Origin::Tuple(scheme, _, _) => {
+                    Ok(Uri {
+                        source: s.to_owned(),
+                        scheme_end: Some(scheme.len()),
+                        authority_end: if authority_end > 0 { Some(authority_end) } else { None },
+                        query: if query_len > 0 { Some(query_len) } else { None },
+                        fragment: if fragment_len > 0 { Some(fragment_len) } else { None },
+                    })
+                }
+            }
+        }
+    }
+
+    pub fn path(&self) -> &str {
+        let index = self.authority_end.unwrap_or(self.scheme_end.unwrap_or(0));
+        let query_len = self.query.unwrap_or(0);
+        let fragment_len = self.fragment.unwrap_or(0);
+        let end = self.source.len() - if query_len > 0 { query_len + 1 } else { 0 } -
+            if fragment_len > 0 { fragment_len + 1 } else { 0 };
+        if index >= end {
+            "/"
+        } else {
+            &self.source[index..end]
+        }
+    }
+
+    pub fn scheme(&self) -> Option<&str> {
+        if let Some(end) = self.scheme_end {
+            Some(&self.source[..end])
+        } else {
+            None
+        }
+    }
+
+    pub fn authority(&self) -> Option<&str> {
+        if let Some(end) = self.authority_end {
+            let index = self.scheme_end.map(|i| i + 3).unwrap_or(0);
+            Some(&self.source[index..end])
+        } else {
+            None
+        }
+    }
+
+    pub fn host(&self) -> Option<&str> {
+        if let Some(auth) = self.authority() {
+            auth.split(":").next()
+        } else {
+            None
+        }
+    }
+
+    pub fn port(&self) -> Option<u16> {
+        if let Some(auth) = self.authority() {
+            let v: Vec<&str> = auth.split(":").collect();
+            if v.len() == 2 {
+                u16::from_str(v[1]).ok()
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn query(&self) -> Option<&str> {
+        let fragment_len = self.fragment.unwrap_or(0);
+        let fragment_len = if fragment_len > 0 { fragment_len + 1 } else { 0 };
+        if let Some(len) = self.query {
+            Some(&self.source[self.source.len() - len - fragment_len..self.source.len() - fragment_len])
+        } else {
+            None
+        }
+    }
+
+    pub fn fragment(&self) -> Option<&str> {
+        if let Some(len) = self.fragment {
+            Some(&self.source[self.source.len() - len..])
+        } else {
+            None
+        }
+    }
+
+    pub fn uri(&self) -> &str {
```

Comment:
I'd probably replace this with an `AsRef<str>` impl, since it sounds like stuttering when saying `req.uri().uri()`.